<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Constrained maximum likelihood (cML) based Mendelian
Randomization method robust to both 
correlated and uncorrelated pleiotropy."><title>Constrained maximum likelihood (cML) method — mr_cML • MendelianRandomization</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Constrained maximum likelihood (cML) method — mr_cML"><meta property="og:description" content="Constrained maximum likelihood (cML) based Mendelian
Randomization method robust to both 
correlated and uncorrelated pleiotropy."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MendelianRandomization</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Data_Input.html">Data Input</a>
    <a class="dropdown-item" href="../articles/Univariable_MR_Methods.html">Univariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Multivariable_MR_Methods.html">Multivariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Graphs_for_MR.html">Graphical summaries of results</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sb452/MendelianRandomizationVignette/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Constrained maximum likelihood (cML) method</h1>
      <small class="dont-index">Source: <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/R/mr_cML-methods.R" class="external-link"><code>R/mr_cML-methods.R</code></a></small>
      <div class="d-none name"><code>mr_cML.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Constrained maximum likelihood (cML) based Mendelian
Randomization method robust to both 
correlated and uncorrelated pleiotropy.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mr_cML</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  MA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  DP <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  K_vec <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">object</span><span class="op">@</span><span class="va">betaX</span><span class="op">)</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  random_start <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  num_pert <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  random_start_pert <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxit <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  random_seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  <span class="va">n</span>,</span>
<span>  Alpha <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for MRInput</span></span>
<span><span class="fu">mr_cML</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  MA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  DP <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  K_vec <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">object</span><span class="op">@</span><span class="va">betaX</span><span class="op">)</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  random_start <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  num_pert <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  random_start_pert <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxit <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  random_seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  <span class="va">n</span>,</span>
<span>  Alpha <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>An <code>MRInput</code> object.</p></dd>


<dt>MA</dt>
<dd><p>Whether model average is applied or not. Default is TRUE.</p></dd>


<dt>DP</dt>
<dd><p>Whether data perturbation is applied or not. Default is TRUE.</p></dd>


<dt>K_vec</dt>
<dd><p>Set of candidate K's, the constraint parameter 
representing number of invalid IVs. Default is from 0 to (#IV - 2).</p></dd>


<dt>random_start</dt>
<dd><p>Number of random starting points for cML, default is 0.</p></dd>


<dt>num_pert</dt>
<dd><p>Number of perturbation when DP is TRUE, default is 200.</p></dd>


<dt>random_start_pert</dt>
<dd><p>Number of random start points for 
cML with data perturbation, default is 0.</p></dd>


<dt>maxit</dt>
<dd><p>Maximum number of iterations for each optimization. 
Default is 100.</p></dd>


<dt>random_seed</dt>
<dd><p>Random seed, default is 314. When <code>random_seed=NULL</code>, no random seed 
will be used and the results may not be reproducible.</p></dd>


<dt>n</dt>
<dd><p>Sample size. When sample sizes of GWAS for exposure and outcome are different,
and/or when sample sizes of different SNPs are different, 
the smallest sample size is recommended to get conservative result and avoid type-I error.
See reference for more discussions.</p></dd>


<dt>Alpha</dt>
<dd><p>Significance level for the confidence interval for estimate, default is 0.05.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The output from the function is an <code>MRcML</code> object containing:</p>
<dl><dt>Exposure</dt>
<dd><p>A character string giving the name given to the exposure.</p></dd>

 <dt>Outcome</dt>
<dd><p>A character string giving the name given to the outcome.</p></dd>

 <dt>Estimate</dt>
<dd><p>Estimate of theta.</p></dd>

 <dt>StdError</dt>
<dd><p>Standard error of estimate.</p></dd>

 <dt>Pvalue</dt>
<dd><p>p-value of estimate.</p></dd>

 <dt>BIC_invalid</dt>
<dd><p>Set of selected invalid IVs if cML-BIC is performed, i.e. without MA or DP.</p></dd>

 <dt>GOF1_p</dt>
<dd><p>p-value of the first goodness-of-fit test.</p></dd>

 <dt>GOF2_p</dt>
<dd><p>p-value of the second goodness-of-fit test.</p></dd>

 <dt>SNPs</dt>
<dd><p>The number of SNPs that were used in the calculation.</p></dd>

 <dt>Alpha</dt>
<dd><p>Significance level for the confidence interval for estimate, default is 0.05.</p></dd>

 <dt>CILower</dt>
<dd><p>Lower bound of the confidence interval for estimate.</p></dd>

 <dt>CIUpper</dt>
<dd><p>Upper bound of the confidence interval for estimate.</p></dd>

 <dt>MA</dt>
<dd><p>Indicator of whether model average is applied.</p></dd>

 <dt>DP</dt>
<dd><p>Indicator of whether data perturbation is applied.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The MRcML method selects invalid IVs with correlated
and/or uncorrelated peliotropic effects using constrained maximum
likelihood. <code>cML-BIC</code> gives results of the selected model with 
original data, while <code>cML-MA-BIC</code> averages over all candidate models.
<code>cML-BIC-DP</code> and <code>cML-MA-BIC-DP</code> are the versions with 
data-perturbation to account for selection uncertainty when 
many invalid IVs have weak pleiotropic effects.</p>
<p>When DP is performed,
two goodness-of-fit (GOF) tests are developed to check whether the 
model-based and DP- based variance estimates converge to the same estimate.
Small p-values of GOF tests indicate selection uncertainty is not ignorable, and 
results from DP is more reliable. See reference for more details.</p>
<p>As the constrained maximum likelihood function is non-convex, multiple
random starting points could be used to find a global minimum. For some 
starting points the algorithm may not converge and a warning message will 
be prompted, typically this will not affect the results.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Xue, H., Shen, X., &amp; Pan, W. (2021). 
Constrained maximum likelihood-based Mendelian randomization 
robust to both correlated and uncorrelated pleiotropic effects. 
The American Journal of Human Genetics, 108(7), 1251-1269.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Perform cML-MA-BIC-DP:</span></span></span>
<span class="r-in"><span><span class="fu">mr_cML</span><span class="op">(</span><span class="fu"><a href="mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>, by <span class="op">=</span> <span class="va">chdlodds</span>,</span></span>
<span class="r-in"><span>byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span>, num_pert<span class="op">=</span><span class="fl">5</span>, MA <span class="op">=</span> <span class="cn">TRUE</span>, DP <span class="op">=</span> <span class="cn">TRUE</span>, n <span class="op">=</span> <span class="fl">17723</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Constrained maximum likelihood method (MRcML) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Variants:  28 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Results for:  cML-MA-BIC-DP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         Method Estimate    SE Pvalue        95% CI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  cML-MA-BIC-DP    3.073 0.278  0.000 [2.528,3.617]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-in"><span><span class="co"># num_pert is set to 5 to reduce computational time</span></span></span>
<span class="r-in"><span><span class="co"># the default value of 200 is recommended in practice</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform cML-BIC-DP:</span></span></span>
<span class="r-in"><span><span class="fu">mr_cML</span><span class="op">(</span><span class="fu"><a href="mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>, by <span class="op">=</span> <span class="va">chdlodds</span>,</span></span>
<span class="r-in"><span>byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span>, MA <span class="op">=</span> <span class="cn">TRUE</span>, DP <span class="op">=</span> <span class="cn">FALSE</span>,, n <span class="op">=</span> <span class="fl">17723</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Constrained maximum likelihood method (MRcML) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Variants:  28 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Results for:  cML-MA-BIC </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Method Estimate    SE Pvalue        95% CI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  cML-MA-BIC    2.895 0.321  0.000 [2.265,3.525]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-in"><span>   </span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Stephen Burgess, Olena Yavorska, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

