<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The mr_mvivw function performs multivariable Mendelian randomization via the inverse-variance method. This is implemented by multivariable weighted linear regression."><title>Multivariable inverse-variance weighted method — mr_mvivw • MendelianRandomization</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariable inverse-variance weighted method — mr_mvivw"><meta property="og:description" content="The mr_mvivw function performs multivariable Mendelian randomization via the inverse-variance method. This is implemented by multivariable weighted linear regression."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MendelianRandomization</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Data_Input.html">Data Input</a>
    <a class="dropdown-item" href="../articles/Univariable_MR_Methods.html">Univariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Multivariable_MR_Methods.html">Multivariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Graphs_for_MR.html">Graphical summaries of results</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sb452/MendelianRandomizationVignette/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Multivariable inverse-variance weighted method</h1>
      <small class="dont-index">Source: <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/R/mr_mvivw-methods.R" class="external-link"><code>R/mr_mvivw-methods.R</code></a></small>
      <div class="d-none name"><code>mr_mvivw.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The <code>mr_mvivw</code> function performs multivariable Mendelian randomization via the inverse-variance method. This is implemented by multivariable weighted linear regression.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mr_mvivw</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  correl.x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nx <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for MRMVInput</span></span>
<span><span class="fu">mr_mvivw</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  correl.x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nx <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>An <code>MRMVInput</code> object.</p></dd>


<dt>model</dt>
<dd><p>What type of model should be used: <code>"default"</code>, <code>"random"</code> or <code>"fixed"</code>. The random-effects model (<code>"random"</code>) is a multiplicative random-effects model, allowing overdispersion in the weighted linear regression (the residual standard error is not fixed to be 1, but is not allowed to take values below 1). The fixed-effect model (<code>"fixed"</code>) sets the residual standard error to be 1. The <code>"default"</code> setting is to use a fixed-effect model with 3 genetic variants or fewer, and otherwise to use a random-effects model.</p></dd>


<dt>robust</dt>
<dd><p>Indicates whether robust regression using the <code>lmrob()</code> function from the package <code>robustbase</code> should be used in the method rather than standard linear regression (<code>lm</code>).</p></dd>


<dt>correl</dt>
<dd><p>If the genetic variants are correlated, then this correlation can be accounted for. The matrix of correlations between must be provided in the <code>MRMVInput</code> object: the elements of this matrix are the correlations between the individual variants (diagonal elements are 1). If a correlation matrix is specified in the <code>MRMVInput</code> object, then <code>correl</code> is set to <code>TRUE</code>.</p></dd>


<dt>correl.x</dt>
<dd><p>Correlation matrix for exposures (Optional). Default is to assume the exposures are uncorrelated. This is only used in the computation of conditional F-statistics.</p></dd>


<dt>nx</dt>
<dd><p>Either a single value, or a vector of sample sizes for the genetic associations with the exposures (one for each exposure, and assumed equal for all variants). If a single value is provided, it is assumed this is the sample size for all exposures. This is optional, and it is only used in the calculation of conditional F statistics. If not supplied, then conditional F statistics are not reported.</p></dd>


<dt>distribution</dt>
<dd><p>The type of distribution used to calculate the confidence intervals. Options are <code>"normal"</code> (default) or <code>"t-dist"</code>.</p></dd>


<dt>alpha</dt>
<dd><p>The significance level used to calculate the confidence interval. The default value is 0.05.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to be passed to the regression method.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The output from the function is an <code>MVIVW</code> object containing:</p>
<dl><dt>Model</dt>
<dd><p>A character string giving the type of model used (<code>"fixed"</code>, <code>"random"</code>, or <code>"default"</code>).</p></dd>

 <dt>Exposure</dt>
<dd><p>A character vector with the names given to the exposure.</p></dd>

 <dt>Outcome</dt>
<dd><p>A character string with the names given to the outcome.</p></dd>

 <dt>Robust</dt>
<dd><p><code>TRUE</code> if robust regression has been used to calculate the estimate, <code>FALSE</code> otherwise.</p></dd>

 <dt>Correlation</dt>
<dd><p>The matrix of genetic correlations.</p></dd>

 <dt>Estimate</dt>
<dd><p>A vector of causal estimates.</p></dd>

 <dt>StdError</dt>
<dd><p>A vector of standard errors of the causal estimates.</p></dd>

 <dt>CILower</dt>
<dd><p>The lower bounds of the causal estimates based on the estimated standard errors and the significance level provided.</p></dd>

 <dt>CIUpper</dt>
<dd><p>The upper bounds of the causal estimates based on the estimated standard errors and the significance level provided.</p></dd>

 <dt>Alpha</dt>
<dd><p>The significance level used when calculating the confidence intervals.</p></dd>

 <dt>Pvalue</dt>
<dd><p>The p-values associated with the estimates (calculated as Estimate/StdError as per Wald test) using a normal or t-distribution (as specified in <code>distribution</code>).</p></dd>

 <dt>SNPs</dt>
<dd><p>The number of genetic variants (SNPs) included in the analysis.</p></dd>

 <dt>RSE</dt>
<dd><p>The estimated residual standard error from the regression model.</p></dd>

 <dt>Heter.Stat</dt>
<dd><p>Heterogeneity statistic (Cochran's Q statistic) and associated p-value: the null hypothesis is that all genetic variants estimate the same causal parameter; rejection of the null is an indication that one or more variants may be pleiotropic.</p></dd>

 <dt>CondFstat</dt>
<dd><p>Conditional F statistics: An approximation of the first-stage conditional F statistics for all variants based on the summarized data. This represents the instrument strength for each exposure conditional on other exposures in the model. This is only reported if the sample sizes for the genetic associations with the exposures are provided.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Multivariable Mendelian randomization is an extension of Mendelian randomization to deal with genetic variants that are associated with multiple risk factors. Two scenarios are envisioned for its use: 1) risk factors that are biologically related, such as lipid fractions; and 2) risk factors where there is potentially a network of causal effects (mediation) from one risk factor to another. In both cases, under the extended assumptions of multivariable Mendelian randomization, coefficients represent the direct causal effects of each risk factor in turn with the other risk factors being fixed.</p>
<p>We implement the method using multivariable weighted linear regression. If the variants are correlated, the method is implemented using generalized weighted linear regression; this is hard coded using matrix algebra.</p>
<p>The causal estimate is obtained by regression of the associations with the outcome on the associations with the risk factors, with the intercept set to zero and weights being the inverse-variances of the associations with the outcome.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Description of approach: Stephen Burgess, Simon G Thompson. Multivariable Mendelian Randomization: the use of pleiotropic genetic variants to estimate causal effects. American Journal of Epidemiology 2015; 181(4):251-260. doi: 10.1093/aje/kwu283.</p>
<p>Description of inverse-variance weighted method: Stephen Burgess, Frank Dudbridge, Simon G Thompson. Re: "Multivariable Mendelian randomization: the use of pleiotropic genetic variants to estimate causal effects." American Journal of Epidemiology 2015; 181(4):290-291. doi: 10.1093/aje/kwv017.</p>
<p>Use for mediation analysis: Stephen Burgess, Deborah J Thompson, Jessica MB Rees, Felix R Day, John R Perry, Ken K Ong. Dissecting causal pathways using Mendelian randomization with summarized genetic data: Application to age at menarche and risk of breast cancer. Genetics 2017; 207(2):481-487. doi: 10.1534/genetics.117.300191.</p>
<p>Calculation of conditional F statistics: Ashish Patel, Dipender Gill, Dmitry Shungin, Christos Mantzoros, Lotte Bjerre Knudsen, Jack Bowden, Stephen Burgess. Robust use of phenotypic heterogeneity at drug target genes for mechanistic insights: application of multivariable Mendelian randomization to <em>GLP1R</em> gene region. Pre-print.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">mr_mvivw</span><span class="op">(</span><span class="fu"><a href="mr_mvinput.html">mr_mvinput</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlc</span>, <span class="va">hdlc</span>, <span class="va">trig</span><span class="op">)</span>, bxse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlcse</span>, <span class="va">hdlcse</span>, <span class="va">trigse</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Multivariable inverse-variance weighted method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (variants uncorrelated, random-effect model)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Variants : 28 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Exposure Estimate Std Error  95% CI       p-value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  exposure_1    1.925     0.439  1.064, 2.786   0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  exposure_2   -0.590     0.555 -1.677, 0.498   0.288</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  exposure_3    0.723     0.230  0.272, 1.174   0.002</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residual standard error =  1.433 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Heterogeneity test statistic = 51.3599 on 25 degrees of freedom, (p-value = 0.0014)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Stephen Burgess, Olena Yavorska, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

