<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MendelianRandomization">
<title>Data Input • MendelianRandomization</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data Input">
<meta property="og:description" content="MendelianRandomization">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MendelianRandomization</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Data_Input.html">Data Input</a>
    <a class="dropdown-item" href="../articles/Univariable_MR_Methods.html">Univariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Multivariable_MR_Methods.html">Multivariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Graphs_for_MR.html">Graphical summaries of results</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sb452/MendelianRandomizationVignette/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Data Input</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/vignettes/Data_Input.Rmd" class="external-link"><code>vignettes/Data_Input.Rmd</code></a></small>
      <div class="d-none name"><code>Data_Input.Rmd</code></div>
    </div>

    
    
<p>The package uses a special class called <em>MRInput</em> within the
analyses in order to pass in all necessary information through one
simple structure rather than inserting the data in parts. In order to
make an <em>MRInput</em> object, one can do the following:</p>
<ul>
<li>specify values for each slot separately, or</li>
<li>extract values from the PhenoScanner web-based database</li>
</ul>
<p>We focus initially on the first option.</p>
<div class="section level2">
<h2 id="the-mrinput-format">The MRInput format<a class="anchor" aria-label="anchor" href="#the-mrinput-format"></a>
</h2>
<p>The <em>MRInput</em> object has the following “slots” :</p>
<ul>
<li>
<em>betaX</em> and <em>betaXse</em> are both numeric vectors
describing the associations of the genetic variants with the exposure.
<em>betaX</em> are the beta-coefficients from univariable regression
analyses of the exposure on each genetic variant in turn, and
<em>betaXse</em> are the standard errors.</li>
<li>
<em>betaY</em> and <em>betaYse</em> are both numeric vectors
describing the associations of the genetic variants with the outcome.
<em>betaY</em> are the beta-coefficients from regression analyses of the
outcome on each genetic variant in turn, and <em>betaYse</em> are the
standard errors.</li>
<li>
<em>correlation</em> is a matrix with the signed correlations
between the variants. If a correlation matrix is not provided, it is
assumed that the variants are uncorrelated.</li>
<li>
<em>exposure</em> is a character string giving the name of the risk
factor, e.g. LDL-cholesterol.</li>
<li>
<em>outcome</em> is a character string giving the name of the
outcome, e.g. coronary heart disease. These inputs are only used in the
graphing functions.</li>
<li>
<em>snps</em> is a character vector of the names of the various
genetic variants (SNPs) in the dataset, e.g. rs12785878. It is not
necessary to name the exposure, outcome, or SNPs, but these names are
used in the graphing functions and may be helpful for keeping track of
various analyses.</li>
</ul>
</div>
<div class="section level2">
<h2 id="create-mrinput-object-manually">Create MRInput object manually<a class="anchor" aria-label="anchor" href="#create-mrinput-object-manually"></a>
</h2>
<p>To generate the <em>MRInput</em> object slot by slot, one can use the
<em>mr_input()</em> function :</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, </span>
<span>                          bxse <span class="op">=</span> <span class="va">ldlcse</span>, </span>
<span>                          by <span class="op">=</span> <span class="va">chdlodds</span>, </span>
<span>                          byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MRInputObject</span>  <span class="co"># example with uncorrelated variants</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP exposure.beta exposure.se outcome.beta outcome.se</span></span>
<span><span class="co">## 1 snp_1         0.026       0.004       0.0677     0.0286</span></span>
<span><span class="co">## 2 snp_2        -0.044       0.004      -0.1625     0.0300</span></span>
<span><span class="co">## 3 snp_3        -0.038       0.004      -0.1054     0.0310</span></span>
<span><span class="co">## 4 snp_4        -0.023       0.003      -0.0619     0.0243</span></span>
<span><span class="co">## 5 snp_5        -0.017       0.003      -0.0834     0.0222</span></span>
<span><span class="co">## 6 snp_6        -0.031       0.006      -0.1278     0.0667</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 22 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, </span>
<span>                              bxse <span class="op">=</span> <span class="va">calciumse</span>, </span>
<span>                              by <span class="op">=</span> <span class="va">fastgluc</span>, </span>
<span>                              byse <span class="op">=</span> <span class="va">fastglucse</span>,</span>
<span>                              corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MRInputObject.cor</span>  <span class="co"># example with correlated variants</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP exposure.beta exposure.se outcome.beta outcome.se</span></span>
<span><span class="co">## 1 snp_1       0.00625     0.00233      0.02805     0.0122</span></span>
<span><span class="co">## 2 snp_2       0.00590     0.00338      0.00953     0.0198</span></span>
<span><span class="co">## 3 snp_3       0.01822     0.00318      0.03646     0.0173</span></span>
<span><span class="co">## 4 snp_4       0.00598     0.00233      0.01049     0.0119</span></span>
<span><span class="co">## 5 snp_5       0.00825     0.00229      0.02357     0.0122</span></span>
<span><span class="co">## 6 snp_6       0.00651     0.00352      0.00204     0.0179</span></span></code></pre>
<p>It is not necessary for all the slots to be filled. For example, some
of the methods do not require <em>bxse</em> to be specified; for
example, the <em>mr_ivw</em> function will still run with <em>bxse</em>
set to zeros. If the vectors <em>bx</em>, <em>bxse</em>, <em>by</em>,
and <em>byse</em> are not of equal length, then an error will be
reported. Note that the package does not implement any harmonization of
associations to the same effect allele; this must be done by the
user.</p>
<p>It is also possible to run the analysis using the syntax:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span><span class="va">ldlc</span>, <span class="va">ldlcse</span>, <span class="va">chdlodds</span>, <span class="va">chdloddsse</span><span class="op">)</span></span></code></pre></div>
<p>However, care must be taken in this case to give the vectors in the
correct order (that is: <em>bx, bxse, by, byse</em>).</p>
</div>
<div class="section level2">
<h2 id="extracting-association-estimates-from-phenoscanner">Extracting association estimates from PhenoScanner<a class="anchor" aria-label="anchor" href="#extracting-association-estimates-from-phenoscanner"></a>
</h2>
<p>The PhenoScanner bioinformatic tool (<a href="http://phenoscanner.medschl.cam.ac.uk" class="external-link uri">http://phenoscanner.medschl.cam.ac.uk</a>) is a curated
database of publicly available results from large-scale genetic
association studies. The database currently contains over 65 billion
associations and association results and over 150 million unique genetic
variants, mostly single nucleotide polymorphisms.</p>
<p>PhenoScanner can be called directly from the MendelianRandomization
package using the <em>pheno_input()</em> function. This creates an
<em>MRInput</em> function, which can be directly used as an input to any
of the estimation functions. For example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mr_ivw.html">mr_ivw</a></span><span class="op">(</span><span class="fu"><a href="../reference/pheno_input.html">pheno_input</a></span><span class="op">(</span>snps<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs12916"</span>, <span class="st">"rs2479409"</span>, </span>
<span>                          <span class="st">"rs217434"</span>, <span class="st">"rs1367117"</span>,</span>
<span>                          <span class="st">"rs4299376"</span>, <span class="st">"rs629301"</span>,</span>
<span>                          <span class="st">"rs4420638"</span>, <span class="st">"rs6511720"</span><span class="op">)</span>,</span>
<span>                   exposure <span class="op">=</span> <span class="st">"Low density lipoprotein"</span>,</span>
<span>                   pmidE <span class="op">=</span> <span class="st">"24097068"</span>, </span>
<span>                   ancestryE <span class="op">=</span> <span class="st">"European"</span>,</span>
<span>                   outcome <span class="op">=</span> <span class="st">"Coronary artery disease"</span>,</span>
<span>                   pmidO <span class="op">=</span> <span class="st">"26343387"</span>,</span>
<span>                   ancestryO <span class="op">=</span> <span class="st">"Mixed"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>(We do not implement this code here, as it requires a connection to
the internet, and hence produces an error if an internet connection
cannot be found. But please copy it and try it for yourself!)</p>
<p>In order to obtain the relevant summary estimates, run the
<em>pheno_input()</em> function with: - <em>snps</em> is a character
vector giving the rsid identifiers of the genetic variants. -
<em>exposure</em> is a character vector giving the name of the risk
factor. - <em>pmidE</em> is the PubMed ID of the paper where the
association estimates with the exposure were first published. -
<em>ancestryE</em> is the ancestry of the participants on whom the
association estimates with the exposure were estimated. (For some traits
and PubMed IDs, results are given for multiple ancestries.) Usually,
ancestry is <em>“European”</em> or <em>“Mixed”</em>. - <em>outcome</em>
is a character vector giving the name of the outcome. - <em>pmidO</em>
is the PubMed ID of the paper where the association estimates with the
outcome were first published. - <em>ancestryO</em> is the ancestry of
the participants on whom the association estimates with the exposure
were estimated.</p>
<p>We note that the spelling of the exposure and outcome, the PubMed ID,
and the ancestry information need to correspond exactly to the values in
the PhenoScanner dataset. If these are not spelled exactly as in the
PhenoScanner dataset (including upper/lower case), the association
estimates will not be found.</p>
</div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>Two sets of data are provided as part of this package:</p>
<ul>
<li>
<em>ldlc, ldlcse, hdlc, hdlcse, trig, trigse, chdlodds,
chdloddsse</em>: these are the associations (beta-coefficients and
standard errors) of 28 genetic variants with LDL-cholesterol,
HDL-cholesterol, triglycerides, and coronary heart disease (CHD) risk
(associations with CHD risk are log odds ratios) taken from Waterworth
et al (2011) “Genetic variants influencing circulating lipid levels and
risk of coronary artery disease”, doi: 10.1161/atvbaha.109.201020.</li>
<li>
<em>calcium, calciumse, fastgluc, fastglucse</em>: these are the
associations (beta-coefficients and standard errors) of 7 genetic
variants in the <em>CASR</em> gene region. These 7 variants are all
correlated, and the correlation matrix is provided as <em>calc.rho</em>.
These data were analysed in Burgess et al (2015) “Using published data
in Mendelian randomization: a blueprint for efficient identification of
causal risk factors”, doi: 10.1007/s10654-015-0011-z.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Stephen Burgess, Olena Yavorska, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
