<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MendelianRandomization">
<title>Multivariable Mendelian randomization • MendelianRandomization</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariable Mendelian randomization">
<meta property="og:description" content="MendelianRandomization">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MendelianRandomization</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Data_Input.html">Data Input</a>
    <a class="dropdown-item" href="../articles/Univariable_MR_Methods.html">Univariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Multivariable_MR_Methods.html">Multivariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Graphs_for_MR.html">Graphical summaries of results</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sb452/MendelianRandomizationVignette/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Multivariable Mendelian randomization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/vignettes/Multivariable_MR_Methods.Rmd" class="external-link"><code>vignettes/Multivariable_MR_Methods.Rmd</code></a></small>
      <div class="d-none name"><code>Multivariable_MR_Methods.Rmd</code></div>
    </div>

    
    
<p>An analytic approach when genetic variants are associated with
multiple risk factors is multivariable Mendelian randomization,
introduced in Burgess and Thompson (2015) “Multivariable Mendelian
randomization: the use of pleiotropic genetic variants to estimate
causal effects”, doi: 10.1093/aje/kwu283.</p>
<p>There are two contexts in which we envision the method being used.
First, when there is a set of related risk factors, such as lipid
fractions. It is difficult to find genetic predictors of HDL-cholesterol
that do not also predict LDL-cholesterol and/or triglycerides.
Multivariable Mendelian randomization allows genetic variants to be
associated with all the risk factors in the model provided that they do
not influence the outcome via other pathways. Secondly, when there is a
network of risk factors, typically a primary risk factor and a secondary
risk factor or mediator. In both cases, multivariable estimates reflect
direct causal effects - the result of intervening on the risk factor
under analysis while keeping other risk factors in the model fixed.</p>
<div class="section level2">
<h2 id="the-mrmvinput-format">the MRMVInput format<a class="anchor" aria-label="anchor" href="#the-mrmvinput-format"></a>
</h2>
<p>As the multivariable method requires genetic associations with
multiple risk factors, a different input function is needed:
<em>mr_mvinput</em>, which creates an <em>MRMVInput</em> object:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVMRInputObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvinput.html">mr_mvinput</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlc</span>, <span class="va">hdlc</span>, <span class="va">trig</span><span class="op">)</span>,</span>
<span>                              bxse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlcse</span>, <span class="va">hdlcse</span>, <span class="va">trigse</span><span class="op">)</span>,</span>
<span>                              by <span class="op">=</span> <span class="va">chdlodds</span>, </span>
<span>                              byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVMRInputObject</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP exposure_1.beta exposure_2.beta exposure_3.beta exposure_1.se</span></span>
<span><span class="co">## 1 snp_1           0.026           0.002           0.016         0.004</span></span>
<span><span class="co">## 2 snp_2          -0.044           0.005          -0.004         0.004</span></span>
<span><span class="co">## 3 snp_3          -0.038           0.003          -0.009         0.004</span></span>
<span><span class="co">##   exposure_2.se exposure_3.se outcome.beta outcome.se</span></span>
<span><span class="co">## 1         0.004         0.008       0.0677     0.0286</span></span>
<span><span class="co">## 2         0.004         0.008      -0.1625     0.0300</span></span>
<span><span class="co">## 3         0.004         0.008      -0.1054     0.0310</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 25 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVMRInputObject.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvinput.html">mr_mvinput</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlc</span>, <span class="va">hdlc</span>, <span class="va">trig</span><span class="op">)</span>,</span>
<span>                                  bxse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ldlcse</span>, <span class="va">hdlcse</span>, <span class="va">trigse</span><span class="op">)</span>,</span>
<span>                                  by <span class="op">=</span> <span class="va">chdlodds</span>, </span>
<span>                                  byse <span class="op">=</span> <span class="va">chdloddsse</span>,</span>
<span>                                  correlation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ldlc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multivariable-inverse-variance-weighted-method">Multivariable inverse-variance weighted method<a class="anchor" aria-label="anchor" href="#multivariable-inverse-variance-weighted-method"></a>
</h2>
<p>The <em>mr_mvivw</em> command performs an extension of the
inverse-variance weighted method for multivariable Mendelian
randomization, implementing multivariable weighted linear regression
(the standard IVW method method performs univariable weighted linear
regression) with the intercept term set to zero. As with the standard
IVW method, a correlation matrix can be specified. Multivariable
Mendelian randomization requires the number of genetic variants to
exceed the number of risk factors; if this is not the case, the method
will return an error.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvivw.html">mr_mvivw</a></span><span class="op">(</span><span class="va">MVMRInputObject</span>, </span>
<span>                         model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>                         correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                         correl.x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                         nx <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                         distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                         alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvivw.html">mr_mvivw</a></span><span class="op">(</span><span class="va">MVMRInputObject</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVIVWObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable inverse-variance weighted method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##  exposure_1    1.925     0.439  1.064, 2.786   0.000</span></span>
<span><span class="co">##  exposure_2   -0.590     0.555 -1.677, 0.498   0.288</span></span>
<span><span class="co">##  exposure_3    0.723     0.230  0.272, 1.174   0.002</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  1.433 </span></span>
<span><span class="co">## Heterogeneity test statistic = 51.3599 on 25 degrees of freedom, (p-value = 0.0014)</span></span></code></pre>
<p><em>Conditional F statistics:</em> The <em>nx</em> option specifies
the sample size(s) for the genetic associations with the exposures
(either a single value if these are all equal, or a vector if they
differ). The <em>correl.x</em> option specifies the correlation between
the genetic associations with the exposures. If the genetic assocations
with the exposures are estimates in separate datasets, then this
correlation matrix should be set to the identity matrix (this is the
default option). This matrix cannot be calculated from summarized data;
if it is unknown, then a sensitivity analysis for its value is
recommended. The <em>nx</em> and <em>correl.x</em> options are not used
in the calculation of estimates from the multivariable inverse-variance
weighted method method, but only in the calculation of conditional F
statistics. Conditional F statistics are the relevant measure of
instrument strength for multivariable Mendelian randomization. For more
detail, see Sanderson, Spiller, Bowden (2021) “Testing and correcting
for weak and pleiotropic instruments in two-sample multivariable
Mendelian randomization”, doi: 10.1002/sim.9133.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVIVWObject.condF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvivw.html">mr_mvivw</a></span><span class="op">(</span><span class="va">MVMRInputObject</span>, nx <span class="op">=</span> <span class="fl">17723</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVIVWObject.condF</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable inverse-variance weighted method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value Cond F-stat</span></span>
<span><span class="co">##  exposure_1    1.925     0.439  1.064, 2.786   0.000        20.3</span></span>
<span><span class="co">##  exposure_2   -0.590     0.555 -1.677, 0.498   0.288        12.9</span></span>
<span><span class="co">##  exposure_3    0.723     0.230  0.272, 1.174   0.002        13.3</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  1.433 </span></span>
<span><span class="co">## Heterogeneity test statistic = 51.3599 on 25 degrees of freedom, (p-value = 0.0014)</span></span></code></pre>
<p>In addition to the multivariable inverse-variance weighted method,
several other multivariable estimation methods are available, which
generally are multivariable extensions of univariable methods:
multivariable MR-Egger, multivariable median-based method, multivariable
lasso method, multivariable constrained maximum likelihood,
multivariable GMM, and multivariable principal components GMM.</p>
</div>
<div class="section level2">
<h2 id="multivariable-mr-egger">Multivariable MR-Egger<a class="anchor" aria-label="anchor" href="#multivariable-mr-egger"></a>
</h2>
<p>Multivariable MR-Egger is an extension of the MR-Egger method to deal
with genetic variants that are associated with multiple risk factors,
introduced in Rees et al (2017) “Extending the MR-Egger method for
multivariable Mendelian randomization to correct for both measured and
unmeasured pleiotropy” DOI: 10.1002/sim.7492</p>
<p>We implement the method using multivariable weighted linear
regression. If the variants are correlated, the method is implemented
using generalized weighted linear regression; this is hard coded using
matrix algebra.</p>
<p>The causal estimate is obtained by regression of the associations
with the outcome on the associations with the risk factors, with the
intercept estimated and weights being the inverse-variances of the
associations with the outcome.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVEggerObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvegger.html">mr_mvegger</a></span><span class="op">(</span><span class="va">MVMRInputObject</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVEggerObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable MR-Egger method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Orientated to exposure : 1 </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##     Exposure Estimate Std Error  95% CI        p-value</span></span>
<span><span class="co">##   exposure_1    2.829     0.515  1.819,  3.839   0.000</span></span>
<span><span class="co">##   exposure_2   -0.663     0.496 -1.636,  0.310   0.182</span></span>
<span><span class="co">##   exposure_3    0.827     0.209  0.417,  1.237   0.000</span></span>
<span><span class="co">##  (intercept)   -0.028     0.010 -0.049, -0.008   0.007</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  1.280 </span></span>
<span><span class="co">## Heterogeneity test statistic = 39.3421 on 24 degrees of freedom, (p-value = 0.0251)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariable-median-based-method">Multivariable median-based method<a class="anchor" aria-label="anchor" href="#multivariable-median-based-method"></a>
</h2>
<p>Multivariable median method is similar to the univariable weighted
median method, except that it is implemented using quantile regression.
This is implemented by multivariable weighted quantile regression, with
the quantile set to 0.5 (the median).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVMedianObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvmedian.html">mr_mvmedian</a></span><span class="op">(</span><span class="va">MVMRInputObject</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVMedianObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable median method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##  exposure_1    2.116     0.482  1.172, 3.061   0.000</span></span>
<span><span class="co">##  exposure_2    0.111     0.656 -1.176, 1.397   0.866</span></span>
<span><span class="co">##  exposure_3    1.039     0.261  0.528, 1.550   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariable-lasso-method">Multivariable lasso method<a class="anchor" aria-label="anchor" href="#multivariable-lasso-method"></a>
</h2>
<p>The mr_mvlasso function performs the multivariable MR-Lasso method as
described in Grant and Burgess (2020) “Pleiotropy robust methods for
multivariable Mendelian randomization” DOI: 10.1002/sim.9156. This
applies lasso-type penalization to the direct effects of genetic
variants on the outcome. The causal estimates are described as
post-lasso estimates, and are obtained by performing the multivariable
IVW method using only those genetic variants that are identified as
valid by the lasso procedure.</p>
<p>Multivariable MR-Lasso extends the multivariable IVW model to include
an intercept term for each genetic variant. These intercept terms
represent associations between the genetic variants and the outcome
which bypass the risk factors. The regularized regression model is
estimated by multivariable weighted linear regression where the
intercept terms are subject to lasso-type penalization. The lasso
penalization will tend to shrink the intercept terms corresponding to
the valid instruments to zero.</p>
<p>The lasso penalty relies on a tuning parameter which controls the
level of sparsity. The default is to use a heterogeneity stopping rule,
but a fixed value may be specified.</p>
<p>As part of the analysis, the genetic variants are orientated so that
all of the associations with one of the risk factors are positive (the
first risk factor is used by default). Re-orientation of the genetic
variants is performed automatically as part of the function.</p>
<p>The MR-Lasso method is performed in two steps. First, a regularized
regression model is fitted, and some genetic variants are identified as
valid instruments. Second, causal effects are estimated using standard
multivariable IVW with only the valid genetic variants. The post-lasso
method will be performed as long as the number of genetic variants
identified as valid instruments is greater than the number of risk
factors. The default heterogeneity stopping rule will always return more
genetic variants as valid instruments than risk factors for
identification. The main estimates given by the method are the
post-lasso estimates. However, parameter estimates from the regularized
regression model used to identify invalid variants are also provided for
completeness.</p>
<p>If a substantial proportion of genetic variants are removed from the
analysis, the multivariable MR-Lasso method may give a false impression
of confidence in the causal estimate due to homogeneity of the
variant-specific causal estimates amongst the remaining variants.
However, it is not reasonable to claim that there is strong evidence for
a causal effect after a large number of variants with heterogeneous
estimates have been removed from the analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVLassoObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvlasso.html">mr_mvlasso</a></span><span class="op">(</span><span class="va">MVMRInputObject</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVLassoObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable MR-Lasso method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Orientated to exposure : 1 </span></span>
<span><span class="co">## Number of variants : 28 </span></span>
<span><span class="co">## Number of valid instruments : 25 </span></span>
<span><span class="co">## Tuning parameter : 0.4641104 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##  exposure_1    1.726     0.383  0.976, 2.477   0.000</span></span>
<span><span class="co">##  exposure_2   -0.126     0.441 -0.991, 0.738   0.775</span></span>
<span><span class="co">##  exposure_3    0.858     0.185  0.496, 1.220   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariable-constrained-maximum-likelihood">Multivariable constrained maximum likelihood<a class="anchor" aria-label="anchor" href="#multivariable-constrained-maximum-likelihood"></a>
</h2>
<p>Multivariable MRcML (MVMRcML) is an extension of MRcML to deal with
multiple exposures, as described in Lin et al (2023) “Robust
multivariable Mendelian randomization based on constrained maximum
likelihood” DOI 10.1016/j.ajhg.2023.02.014 . It is robust to both
correlated and uncorrelated pleiotropy as with the univariable
equivalent.</p>
<p>In practice, the data perturbation (DP) version is preferred in
practice for a more robust inference as it can account for the
uncertainty in model selection. However, it may take a longer time
especially when the number of IVs is large (so the range of K_vec can be
large too). One strategy is to try a small range of K (the number of
invalid IVs) first (with a small num_pert), then adjust it if the number
of selected invalid IVs fall close to the boundary. You can also use
other methods, e.g. mr_mvlasso, to get a rough sense of the number of
invalid IVs.</p>
<p>Similar to mr_cML, multiple random starting points could be used to
find a global minimum.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVcMLObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvcML.html">mr_mvcML</a></span><span class="op">(</span><span class="va">MVMRInputObject</span>, n <span class="op">=</span> <span class="fl">17723</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVcMLObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable MRcML method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of variants : 28 </span></span>
<span><span class="co">## Results for:  MVMRcML-DP </span></span>
<span><span class="co">## Number of data perturbations with successful convergence:  200 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##  exposure_1    1.927     0.613  0.726, 3.127   0.002</span></span>
<span><span class="co">##  exposure_2   -0.350     0.671 -1.666, 0.966   0.602</span></span>
<span><span class="co">##  exposure_3    0.822     0.272  0.289, 1.354   0.002</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariable-gmm">Multivariable GMM<a class="anchor" aria-label="anchor" href="#multivariable-gmm"></a>
</h2>
<p>Multivariable Mendelian randomization via the generalized method of
moments method as described in Hansen (1982) “Large sample properties of
generalized method of moments estimators” DOI 10.2307/1912775</p>
<p>The method accounts for overdispersion heterogeneity in genetic
variant-outcome associations.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVGMMObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvgmm.html">mr_mvgmm</a></span><span class="op">(</span><span class="va">MVMRInputObject</span>, nx<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">17723</span>,<span class="fl">3</span><span class="op">)</span>, ny<span class="op">=</span><span class="fl">17723</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVGMMObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable generalized method of moments (GMM) method</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Exposure correlation matrix not specified. Exposures are assumed to be uncorrelated.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Robust model with overdispersion heterogeneity.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value Cond F-stat</span></span>
<span><span class="co">##  exposure_1    1.830     0.451  0.947, 2.713   0.000        20.3</span></span>
<span><span class="co">##  exposure_2   -0.722     0.580 -1.858, 0.415   0.213        12.9</span></span>
<span><span class="co">##  exposure_3    0.686     0.239  0.217, 1.155   0.004        13.3</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Overdispersion heterogeneity parameter estimate = 17.369 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Heterogeneity test statistic = 24.9273</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariable-principal-components-gmm">Multivariable principal components GMM<a class="anchor" aria-label="anchor" href="#multivariable-principal-components-gmm"></a>
</h2>
<p>Multivariable principal components GMM is a multivariable version of
the PC-GMM method, as described in Patel et al. “Robust use of
phenotypic heterogeneity at drug target genes for mechanistic insights:
application of cis-multivariable Mendelian randomization to GLP1R gene
region” DOI 10.1101/2023.07.20.23292958</p>
<p>This can be used when there are distinct exposures associated with
variants at a single gene region. Phenotypic heterogeneity (that is, the
genetic associations with the exposures are not collinear) at genomic
loci encoding drug targets can be exploited by multivariable Mendelian
randomization to provide insight on the pathways by which
pharmacological interventions may affect disease risk.</p>
<p>When a Mendelian randomization analysis is performed using correlated
genetic variants from a single gene region, there is a tradeoff between
using too few variants (and compromising on power) and using too many
variants (in which case, estimates can be highly sensitive to small
variation in the correlation matrix). This method performs dimension
reduction on a weighted version of the genetic correlation matrix to
form principal components based on the genetic variants, which are then
used as instruments. It is recommended not to include very highly
correlated variants in this method (say, r^2 &gt; 0.95), but the method
should cope well with variants correlated below this level.</p>
<p>This method provides two-sample multivariable Mendelian randomization
estimates and associated confidence intervals that account for
overdispersion heterogeneity in dimension-reduced genetic associations
(when robust = TRUE).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MVpcGMMObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mvpcgmm.html">mr_mvpcgmm</a></span><span class="op">(</span><span class="va">MVMRInputObject.cor</span>, nx<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">17723</span>,<span class="fl">3</span><span class="op">)</span>, ny<span class="op">=</span><span class="fl">17723</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MVpcGMMObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multivariable principal components generalized method of moments (PC-GMM) method</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Exposure correlation matrix not specified. Exposures are assumed to be uncorrelated.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of principal components used : 25 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Robust model with overdispersion heterogeneity.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##    Exposure Estimate Std Error  95% CI       p-value Cond F-stat</span></span>
<span><span class="co">##  exposure_1    1.695     0.517  0.681, 2.709   0.001        18.6</span></span>
<span><span class="co">##  exposure_2   -0.754     0.759 -2.240, 0.733   0.321        10.2</span></span>
<span><span class="co">##  exposure_3    0.682     0.292  0.110, 1.255   0.020        10.8</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Overdispersion heterogeneity parameter estimate = 25.5785 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Heterogeneity test statistic = 21.5116</span></span></code></pre>
<p>Note that this <em>mr_mvpcgmm</em> example does not use variants from
a single gene region. It is provided to demonstrate that the code works,
rather than to illustrate a recommended use case.</p>
</div>
<div class="section level2">
<h2 id="final-note-of-caution">Final note of caution<a class="anchor" aria-label="anchor" href="#final-note-of-caution"></a>
</h2>
<p>Particularly with the development of Mendelian randomization with
summarized data, two-sample Mendelian randomization (where associations
with the risk factor and with the outcome are taken from separate
datasets), and bioinformatic tools for obtaining and analysing
summarized data (including this package), Mendelian randomization is
becoming increasingly accessible as a tool for answering epidemiological
questions. This is undoubtably a Good Thing.</p>
<p>However, it is important to remember that the difficult part of a
Mendelian randomization analysis is not the computational method, but
deciding what should go into the analysis: which risk factor, which
outcome, and (most importantly) which genetic variants. Hopefully, the
availability of these tools will enable less attention to be paid to the
mechanics of the analysis, and more attention to these choices.</p>
<p>The note of caution is that tools that make Mendelian randomization
simple to perform run the risk of encouraging large numbers of
speculative analyses to be performed in an unprincipled way. It is
important that Mendelian randomization is not performed in a way that
avoids critical thought. In releasing this package, the hope is that it
will lead to more comprehensive and more reproducible causal inferences
from Mendelian randomization, and not simply add more noise to the
literature.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Stephen Burgess, Olena Yavorska, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
