<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MendelianRandomization">
<title>Univariable estimation methods • MendelianRandomization</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Univariable estimation methods">
<meta property="og:description" content="MendelianRandomization">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MendelianRandomization</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Data_Input.html">Data Input</a>
    <a class="dropdown-item" href="../articles/Univariable_MR_Methods.html">Univariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Multivariable_MR_Methods.html">Multivariable estimation methods</a>
    <a class="dropdown-item" href="../articles/Graphs_for_MR.html">Graphical summaries of results</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sb452/MendelianRandomizationVignette/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Univariable estimation methods</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sb452/MendelianRandomizationVignette/blob/HEAD/vignettes/Univariable_MR_Methods.Rmd" class="external-link"><code>vignettes/Univariable_MR_Methods.Rmd</code></a></small>
      <div class="d-none name"><code>Univariable_MR_Methods.Rmd</code></div>
    </div>

    
    
<p>The MendelianRandomization package supports several univariable
estimation methods (that is, methods for a single exposure variable):
the inverse-variance weighted method, the median-based method, the
MR-Egger method, the mode-based method, the maximum likelihood method,
the heterogeneity penalized method, the contamination mixture method,
the lasso method, the debiased inverse-variance weighted method, the
penalized inverse-variance weighted method, the constrained maximum
likelihood method, and the principal components generalized method of
moments (GMM) method. We describe these methods in turn.</p>
<div class="section level3">
<h3 id="inverse-variance-weighted-method">Inverse-variance weighted method<a class="anchor" aria-label="anchor" href="#inverse-variance-weighted-method"></a>
</h3>
<p>The inverse-variance method is equivalent to the standard
instrumental variable method using individual-level data, the two-stage
least squares method. Either a fixed- or a random-effects analysis can
be performed; the <em>“default”</em> option is a fixed-effect analysis
when there are three variants or fewer, and a random-effects analysis
otherwise. The <em>robust</em> option uses robust regression rather than
standard regression in the analysis, and the <em>penalized</em> option
downweights the contribution to the analysis of genetic variants with
outlying (heterogeneous) variant-specific estimates. If a correlation
matrix is provided in the <em>MRInput</em> object, then the correlated
method is used by default (<em>correl = TRUE</em>), and the
<em>robust</em> and <em>penalized</em> arguments are ignored.</p>
<p>The default options for constructing confidence intervals are based
on a normal distribution and a 95% confidence level, however one can use
the t-distribution (<em>distribution = “t-dist”</em>) and alternative
significance level if desired.</p>
<p>The default option for the weights is <em>“simple”</em>, which
corresponds to an inverse-variance weighted meta-analysis of the ratio
estimates <em>by/bx</em> using first-order standard errors
<em>byse/bx</em>. Alternatively, weights can be set to <em>“delta”</em>,
to include second-order terms of the delta expansion. This makes use of
<em>psi</em>, which is the correlation for each variant between its
association with the exposure and with the outcome.</p>
<p>The <em>mr_ivw</em> function automatically calculates an
approximation to the first-stage F statistic from regression of the
exposure on the genetic variants. This is an important measure of
instrument strength. When the correlation matrix is specified,
calculation of the estimated F statistic accounts for the correlation
between variants. When a correlation matrix is specified, a warning
message appears to remind the user of the importance of correct
harmonization of the correlation matrix to the same effect and other
alleles as the genetic associations with the exposure and outcome.</p>
<div class="section level4">
<h4 id="ivw-example-with-uncorrelated-variants">IVW Example with uncorrelated variants<a class="anchor" aria-label="anchor" href="#ivw-example-with-uncorrelated-variants"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, </span>
<span>                          bxse <span class="op">=</span> <span class="va">ldlcse</span>, </span>
<span>                          by <span class="op">=</span> <span class="va">chdlodds</span>, </span>
<span>                          byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MRInputObject</span>  <span class="co"># example with uncorrelated variants</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP exposure.beta exposure.se outcome.beta outcome.se</span></span>
<span><span class="co">## 1 snp_1         0.026       0.004       0.0677     0.0286</span></span>
<span><span class="co">## 2 snp_2        -0.044       0.004      -0.1625     0.0300</span></span>
<span><span class="co">## 3 snp_3        -0.038       0.004      -0.1054     0.0310</span></span>
<span><span class="co">## 4 snp_4        -0.023       0.003      -0.0619     0.0243</span></span>
<span><span class="co">## 5 snp_5        -0.017       0.003      -0.0834     0.0222</span></span>
<span><span class="co">## 6 snp_6        -0.031       0.006      -0.1278     0.0667</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 22 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_ivw.html">mr_ivw</a></span><span class="op">(</span><span class="va">MRInputObject</span>,</span>
<span>                    model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>                    robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                    penalized <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                    correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                    weights <span class="op">=</span> <span class="st">"simple"</span>,</span>
<span>                    psi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                    distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                    alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">IVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_ivw.html">mr_ivw</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>   by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">IVWObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Inverse-variance weighted method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##     IVW    2.834     0.530 1.796, 3.873   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  1.920 </span></span>
<span><span class="co">## Heterogeneity test statistic (Cochran's Q) = 99.5304 on 27 degrees of freedom, (p-value = 0.0000). I^2 = 72.9%. </span></span>
<span><span class="co">## F statistic = 28.0.</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, </span>
<span>                              bxse <span class="op">=</span> <span class="va">calciumse</span>, </span>
<span>                              by <span class="op">=</span> <span class="va">fastgluc</span>, </span>
<span>                              byse <span class="op">=</span> <span class="va">fastglucse</span>,</span>
<span>                              corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="example-with-uncorrelated-variants">Example with uncorrelated variants<a class="anchor" aria-label="anchor" href="#example-with-uncorrelated-variants"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRInputObject.cor</span>  <span class="co"># example with correlated variants</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP exposure.beta exposure.se outcome.beta outcome.se</span></span>
<span><span class="co">## 1 snp_1       0.00625     0.00233      0.02805     0.0122</span></span>
<span><span class="co">## 2 snp_2       0.00590     0.00338      0.00953     0.0198</span></span>
<span><span class="co">## 3 snp_3       0.01822     0.00318      0.03646     0.0173</span></span>
<span><span class="co">## 4 snp_4       0.00598     0.00233      0.01049     0.0119</span></span>
<span><span class="co">## 5 snp_5       0.00825     0.00229      0.02357     0.0122</span></span>
<span><span class="co">## 6 snp_6       0.00651     0.00352      0.00204     0.0179</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IVWObject.correl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_ivw.html">mr_ivw</a></span><span class="op">(</span><span class="va">MRInputObject.cor</span>,</span>
<span>                    model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>                    correl <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                    alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">IVWObject.correl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_ivw.html">mr_ivw</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, bxse <span class="op">=</span> <span class="va">calciumse</span>,</span>
<span>   by <span class="op">=</span> <span class="va">fastgluc</span>, byse <span class="op">=</span> <span class="va">fastglucse</span>, corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">IVWObject.correl</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Inverse-variance weighted method</span></span>
<span><span class="co">## (variants correlated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##     IVW    2.245     0.643 0.984, 3.505   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  0.641 </span></span>
<span><span class="co">## Residual standard error is set to 1 in calculation of confidence interval when its estimate is less than 1.</span></span>
<span><span class="co">## Heterogeneity test statistic (Cochran's Q) = 2.0530 on 5 degrees of freedom, (p-value = 0.8418). I^2 = 0.0%. </span></span>
<span><span class="co">## F statistic = 11.5. </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Estimates with correlated variants are sensitive to the signs in the correlation matrix</span></span>
<span><span class="co">##  - please ensure that your correlations are expressed with respect to the same effect alleles as your summarized association estimates.)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="median-based-method">Median-based method<a class="anchor" aria-label="anchor" href="#median-based-method"></a>
</h3>
<p>The median-based method calculates a median of the variant-specific
estimates from the ratio method for each genetic variant individually.
The default option is to calculate a weighted median using the
inverse-variance weights. Alternatively, one can calculate a simple
(unweighted) median, or a weighted median using penalization of weights
for heterogeneous variants. Since the calculation of standard error
requires bootstrapping, the number of bootstrap iterations can be
varied. The random seed is set automatically so that results are
reproducible; however, the value of the seed can be changed if required.
Once the function has been completed, the random seed is reset to its
previous value (this is true for all functions in the package).</p>
<p>The median-based method requires data on at least 3 genetic variants.
Variants must be uncorrelated.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">WeightedMedianObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_median.html">mr_median</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                                  weighting <span class="op">=</span> <span class="st">"weighted"</span>, </span>
<span>                                  distribution <span class="op">=</span> <span class="st">"normal"</span>, </span>
<span>                                  alpha <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                                  iterations <span class="op">=</span> <span class="fl">10000</span>, </span>
<span>                                  seed <span class="op">=</span> <span class="fl">314159265</span><span class="op">)</span></span>
<span></span>
<span><span class="va">WeightedMedianObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_median.html">mr_median</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">WeightedMedianObject</span> </span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Weighted median method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##                  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##  Weighted median method    2.683     0.419 1.862, 3.504   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SimpleMedianObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_median.html">mr_median</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span>, weighting <span class="op">=</span> <span class="st">"simple"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SimpleMedianObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Simple median method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##                Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##  Simple median method    1.755     0.740 0.305, 3.205   0.018</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mr-egger-method">MR-Egger method<a class="anchor" aria-label="anchor" href="#mr-egger-method"></a>
</h3>
<p>The MR-Egger method is implemented here using a random-effects model
only. The <em>robust</em> and <em>penalized</em> options are the same as
for the inverse-variance weighted method. The method can be used for
both correlated and uncorrelated sets of variants. Confidence intervals
can be constructed either using a normal distribution (<em>distribution
= “normal”</em>, the default option), or a t-distribution
(<em>distribution = “t-dist”</em>).</p>
<p>With a t-distribution, in case of under-dispersion (the estimated
residual standard error in the regression model is less than 1),
confidence intervals and p-values use either a t-distribution with no
correction for under-dispersion, or a normal distribution with the
residual standard error set to 1 – whichever is wider. This means that
under-dispersion is not doubly penalized by setting the residual
standard error to 1 and using a t-distribution, but also that the
confidence intervals are not narrower (p-values not more extreme) than
those using a fixed-effect model.</p>
<p>The MR-Egger method requires data on at least 3 genetic variants.
Variants are permitted to be correlated.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EggerObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_egger.html">mr_egger</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                        robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        penalized <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EggerObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_egger.html">mr_egger</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EggerObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## MR-Egger method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants =  28 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##       Method Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##     MR-Egger    3.253     0.770  1.743, 4.762   0.000</span></span>
<span><span class="co">##  (intercept)   -0.011     0.015 -0.041, 0.018   0.451</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual Standard Error :  1.935 </span></span>
<span><span class="co">## Heterogeneity test statistic = 97.3975 on 26 degrees of freedom, (p-value = 0.0000)</span></span>
<span><span class="co">## I^2_GX statistic: 91.9%</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EggerObject.corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_egger.html">mr_egger</a></span><span class="op">(</span><span class="va">MRInputObject.cor</span>, </span>
<span>                        correl <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EggerObject.corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_egger.html">mr_egger</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, bxse <span class="op">=</span> <span class="va">calciumse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">fastgluc</span>, byse <span class="op">=</span> <span class="va">fastglucse</span>, corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EggerObject.corr</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## MR-Egger method</span></span>
<span><span class="co">## (variants correlated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants =  6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##       Method Estimate Std Error  95% CI       p-value</span></span>
<span><span class="co">##     MR-Egger    1.302     1.518 -1.672, 4.276   0.391</span></span>
<span><span class="co">##  (intercept)    0.009     0.013 -0.017, 0.035   0.493</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual Standard Error :  0.629 </span></span>
<span><span class="co">## Residual standard error is set to 1 in calculation of confidence interval when its estimate is less than 1.</span></span>
<span><span class="co">## Heterogeneity test statistic = 1.5825 on 4 degrees of freedom, (p-value = 0.8119)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="maximum-likelihood-method">Maximum likelihood method<a class="anchor" aria-label="anchor" href="#maximum-likelihood-method"></a>
</h3>
<p>An alternative estimation method is the maximum likelihood method,
introduced in Burgess et al (2013) “Mendelian randomization analysis
with multiple genetic variants using summarized data”, doi:
10.1002/gepi.21758.</p>
<p>The method has two main advantages over the IVW method: it allows for
uncertainty in the genetic associations with the exposure (which is
ignored in the IVW method using simple weights), and it allows for
genetic associations with the exposure and with the outcome for each
variant to be correlated. This correlation arises if the samples for the
associations with the exposure and the outcome overlap. In a strict
two-sample Mendelian randomization setting, the correlation parameter
<em>psi</em> is set to zero (the default option). If the associations
are estimated in the same individuals (complete sample overlap), then
the observed correlation between the exposure and the outcome is a
reasonable proposal for the value of <em>psi</em>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MaxLikObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_maxlik.html">mr_maxlik</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                          model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>                          correl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          psi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                          distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                          alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MaxLikObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_maxlik.html">mr_maxlik</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MaxLikObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Maximum-likelihood method</span></span>
<span><span class="co">## (variants uncorrelated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##  MaxLik    3.225     0.569 2.110, 4.340   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  1.793 </span></span>
<span><span class="co">## Heterogeneity test statistic = 86.8145 on 27 degrees of freedom, (p-value = 0.0000)</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MaxLikObject.corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_maxlik.html">mr_maxlik</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, bxse <span class="op">=</span> <span class="va">calciumse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">fastgluc</span>, byse <span class="op">=</span> <span class="va">fastglucse</span>, corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MaxLikObject.corr</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Maximum-likelihood method</span></span>
<span><span class="co">## (variants correlated, random-effect model)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 6 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##  MaxLik    2.303     0.709 0.914, 3.692   0.001</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Residual standard error =  0.588 </span></span>
<span><span class="co">## Residual standard error is set to 1 in calculation of confidence interval when its estimate is less than 1.</span></span>
<span><span class="co">## Heterogeneity test statistic = 1.7277 on 5 degrees of freedom, (p-value = 0.8854)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mode-based-estimation-method">Mode-based estimation method<a class="anchor" aria-label="anchor" href="#mode-based-estimation-method"></a>
</h3>
<p>An additional method for robust estimation that gives consistent when
a plurality of genetic variants is valid is the mode-based estimation
method, introduced in Hartwig, Davey Smith and Bowden (2017) “Robust
inference in summary data Mendelian randomization via the zero modal
pleiotropy assumption”, doi: 10.1093/ije/dyx102.</p>
<p>While the median-based method calculates the variant-specific
estimates and then obtains the median of these estimates, and the
inverse-variance weighted estimate is a weighted mean of the
variant-specific estimates, the mode-based method obtains the “mode” of
these estimates. However, in finite samples, no two estimates will be
exactly the same, so a mode does not exist. The method proceeds by
constructing a kernel-weighted density of the variant-specific
estimates, and taking the maximum point of this density as the point
estimate. A confidence interval is obtained by bootstrapping.</p>
<p>Several options can be specified: whether the mode should be
<em>“weighted”</em> or <em>“unweighted”</em> (default is weighted),
whether the standard errors should be calculated using the
<em>“simple”</em> or <em>“delta”</em> formula (default is
<em>delta</em>), the value of the bandwidth parameter multiplication
factor (default is 1), the random seed, and number of iterations in the
bootstrap procedure.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MBEObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mbe.html">mr_mbe</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                    weighting <span class="op">=</span> <span class="st">"weighted"</span>,</span>
<span>                    stderror <span class="op">=</span> <span class="st">"delta"</span>,</span>
<span>                    phi <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    seed <span class="op">=</span> <span class="fl">314159265</span>,</span>
<span>                    iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                    distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                    alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MBEObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_mbe.html">mr_mbe</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MBEObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Mode-based method of Hartwig et al</span></span>
<span><span class="co">## (weighted, delta standard errors [not assuming NOME], bandwidth factor = 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##     MBE    2.946     1.005 0.977, 4.916   0.003</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="heterogeneity-penalized-method">Heterogeneity-penalized method<a class="anchor" aria-label="anchor" href="#heterogeneity-penalized-method"></a>
</h3>
<p>Another method for robust estimation that also gives consistent when
a plurality of genetic variants is valid is the heterogeneity-penalized
model-averaging method, introduced in Burgess et al (2018) “Modal-based
estimation via heterogeneity-penalized weighting: model averaging for
consistent and efficient estimation in Mendelian randomization when a
plurality of candidate instruments are valid”, doi:
10.1093/ije/dyy08010.1101/175372.</p>
<p>The heterogeneity-penalized method uses the same consistency
criterion as the mode-based estimation method, but evaluates the modal
estimate in a different way. It proceeds by evaluating weights for all
subsets of genetic variants (excluding the null set and singletons).
Subsets receive greater weight if they include more variants, but are
severely downweighted if the variants in the subset have heterogeneous
variant-specific estimates. As such, the method will identify the subset
with the largest number (by weight) of variants having similar
variant-specific estimates.</p>
<p>Confidence intervals are evaluated by calculating a log-likelihood
function, and finding all points within a given vertical distance of the
maximum of the log-likelihood function (which is the overall estimate).
As such, if the log-likelihood function is multimodal, then the
confidence interval may include multiple disjoint ranges. This may
indicate the presence of multiple causal mechanisms by which the
exposure may influence the outcome with different magnitudes of causal
effect. As the confidence interval is determined by a grid search, care
must be taken when chosing the minimum (<em>CIMin</em>) and maximum
(<em>CIMax</em>) values in the search, as well as the step size
(<em>CIStep</em>). The default values will not be suitable for all
applications.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HetPenObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_hetpen.html">mr_hetpen</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                          prior <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                          CIMin <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>                          CIMax <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          CIStep <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>                          alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>As the method evaluates estimates and weights for each subset of
variants, the method is quite computationally expensive to run, as the
complexity doubles with each additional variant. We run the method for a
subset of 10 genetic variants:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HetPenObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_hetpen.html">mr_hetpen</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>, CIMin <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, CIMax <span class="op">=</span> <span class="fl">5</span>, CIStep <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span></span>
<span><span class="va">HetPenObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Heterogeneity-penalized method</span></span>
<span><span class="co">## (Prior probability of instrument validity = 0.5)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 10 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate 95% CI     </span></span>
<span><span class="co">##  HetPen     2.85  1.87, 3.90</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Note: confidence interval is a single range of values.</span></span></code></pre>
<p>As an example of a multimodal confidence interval:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcrp</span>    <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.160</span>, <span class="fl">0.236</span>, <span class="fl">0.149</span>, <span class="fl">0.09</span>, <span class="fl">0.079</span>, <span class="fl">0.072</span>, <span class="fl">0.047</span>, <span class="fl">0.069</span><span class="op">)</span></span>
<span><span class="va">bcrpse</span>  <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.006</span>, <span class="fl">0.009</span>, <span class="fl">0.006</span>, <span class="fl">0.005</span>, <span class="fl">0.005</span>, <span class="fl">0.005</span>, <span class="fl">0.006</span>, <span class="fl">0.011</span><span class="op">)</span></span>
<span><span class="va">bchd</span>    <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0237903</span>, <span class="op">-</span><span class="fl">0.1121942</span>, <span class="op">-</span><span class="fl">0.0711906</span>, <span class="op">-</span><span class="fl">0.030848</span>, <span class="fl">0.0479207</span>, <span class="fl">0.0238895</span>,</span>
<span>  <span class="fl">0.005528</span>, <span class="fl">0.0214852</span><span class="op">)</span></span>
<span><span class="va">bchdse</span>  <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0149064</span>, <span class="fl">0.0303084</span>, <span class="fl">0.0150552</span>, <span class="fl">0.0148339</span>, <span class="fl">0.0143077</span>, <span class="fl">0.0145478</span>,</span>
<span>  <span class="fl">0.0160765</span>, <span class="fl">0.0255237</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">HetPenObject.multimode</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_hetpen.html">mr_hetpen</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">bcrp</span>, bxse <span class="op">=</span> <span class="va">bcrpse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">bchd</span>, byse <span class="op">=</span> <span class="va">bchdse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">HetPenObject.multimode</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Heterogeneity-penalized method</span></span>
<span><span class="co">## (Prior probability of instrument validity = 0.5)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 8 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate  95% CI       </span></span>
<span><span class="co">##  HetPen   -0.437 -0.599, -0.256</span></span>
<span><span class="co">##                   0.014,  0.449</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Note: confidence interval contains multiple ranges of values.</span></span></code></pre>
<p>Several other estimation methods are available, which we do not
describe in detail, as their operation is similar to methods already
described above. We would encourage interested analysts to read the help
files for each method, which describe the inputs and outputs of each
method.</p>
</div>
<div class="section level3">
<h3 id="contamination-mixture-method">Contamination mixture method<a class="anchor" aria-label="anchor" href="#contamination-mixture-method"></a>
</h3>
<p>The contamination mixture method is introduced in Burgess et al
(2020) “A robust and efficient method for Mendelian randomization with
hundreds of genetic variants”, doi: 10.1038/s41467-019-14156-4. It is
similar in spirit to the heterogeneity-penalized method, but much more
computationally efficient.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ConMixObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_conmix.html">mr_conmix</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                          psi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                          CIMin <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                          CIMax <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                          CIStep <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                          alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ConMixObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_conmix.html">mr_conmix</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ConMixObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Contamination mixture method</span></span>
<span><span class="co">## (Standard deviation of invalid estimands = 66.17905)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate 95% CI       p-value</span></span>
<span><span class="co">##  ConMix     2.73  2.00, 3.43 4.79e-15</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## Note: confidence interval is a single range of values.</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="lasso-method">Lasso method<a class="anchor" aria-label="anchor" href="#lasso-method"></a>
</h2>
<p>The lasso method is introduced in Rees et al (2019) “Robust methods
in Mendelian randomization via penalization of heterogeneous causal
estimates”, doi: 10.1371/journal.pone.0222362. It is an outlier-robust
method, which excludes variants with heterogeneous variant-specific
estimates using lasso penalization.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LassoObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_lasso.html">mr_lasso</a></span><span class="op">(</span><span class="va">MRInputObject</span>, </span>
<span>                        distribution <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                        lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LassoObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_lasso.html">mr_lasso</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LassoObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## MR-Lasso method </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of variants : 28 </span></span>
<span><span class="co">## Number of valid instruments : 26 </span></span>
<span><span class="co">## Tuning parameter : 0.7757552 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Exposure Estimate Std Error 95% CI       p-value</span></span>
<span><span class="co">##  exposure    2.671     0.431 1.827, 3.515   0.000</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
<div class="section level3">
<h3 id="debiased-inverse-variance-weighted-method">Debiased inverse-variance weighted method<a class="anchor" aria-label="anchor" href="#debiased-inverse-variance-weighted-method"></a>
</h3>
<p>The debiased inverse-variance weighted method is a variation on the
inverse-variance weighted method that is more robust to bias due to weak
instruments and winner’s curse. It is described in Ting, Shao, Kang
(2021) “Debiased inverse-variance weighted estimator in two-sample
summary-data Mendelian randomization”, doi: 10.1214/20-aos2027.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_divw.html">mr_divw</a></span><span class="op">(</span><span class="va">MRInputObject</span>,</span>
<span>                      over.dispersion <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                      diagnostics <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">DIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_divw.html">mr_divw</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">DIVWObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Debiased inverse-variance weighted method</span></span>
<span><span class="co">## (Over.dispersion:TRUE)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value Condition</span></span>
<span><span class="co">##    dIVW    2.940     0.531 1.900, 3.980   0.000   142.920</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="penalized-inverse-variance-weighted-method">Penalized inverse-variance weighted method<a class="anchor" aria-label="anchor" href="#penalized-inverse-variance-weighted-method"></a>
</h3>
<p>The penalized inverse-variance weighted method is a further extension
of the inverse-variance weighted and debiased inverse-variance weighted
methods to deal with selection of variants more explicitly. It is
described in Wang et al (2022) “A novel penalized inverse-variance
weighted estimator for Mendelian randomization with applications to
COVID-19 outcomes”, doi: 10.1111/biom.13732.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_pivw.html">mr_pivw</a></span><span class="op">(</span><span class="va">MRInputObject</span>,</span>
<span>                      over.dispersion <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      delta <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                      sel.pval <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      Boot.Fieller <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">PIVWObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_pivw.html">mr_pivw</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">PIVWObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Penalized inverse-variance weighted method</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Over dispersion: TRUE </span></span>
<span><span class="co">## Bootstrapping Fieller: TRUE </span></span>
<span><span class="co">## Penalty parameter (lambda): 1 </span></span>
<span><span class="co">## IV selection threshold (delta): 0 </span></span>
<span><span class="co">## Number of variants : 28 </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value Condition</span></span>
<span><span class="co">##    pIVW    2.932     0.528 1.922, 4.001   0.000   142.920</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="constrained-maximum-likelihood-method">Constrained maximum likelihood method<a class="anchor" aria-label="anchor" href="#constrained-maximum-likelihood-method"></a>
</h3>
<p>The constrained maximum likelihood method is a robust method that can
account for violation of any of the three instrumental variable
assumptions under mild assumptions. In a maximum likelihood framework,
this method constrains the number of invalid instruments with horizontal
pleiotropy. The number of invalid instruments is asymptotically
consistently selected by the Bayesian information criterion. It is
described in Xue, Shen, Pan (2021) “Constrained maximum likelihood-based
Mendelian randomization robust to both correlated and uncorrelated
pleiotropic effects”, doi: 10.1016/j.ajhg.2021.05.014.</p>
<p>Note that this method requires the sample size used to calculate the
genetic associations with the exposure and/or outcome to be specified
(if these differ, then it is recommended to use the lower value; see
reference for details).</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cMLObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_cML.html">mr_cML</a></span><span class="op">(</span><span class="va">MRInputObject</span>,</span>
<span>                    MA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    DP <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    K_vec <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">object</span><span class="op">@</span><span class="va">betaX</span><span class="op">)</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span>, </span>
<span>                    random_start <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                    num_pert <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                    random_start_pert <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                    maxit <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                    random_seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>                    <span class="va">n</span>,</span>
<span>                    Alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cMLObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_cML.html">mr_cML</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">ldlc</span>, bxse <span class="op">=</span> <span class="va">ldlcse</span>,</span>
<span>  by <span class="op">=</span> <span class="va">chdlodds</span>, byse <span class="op">=</span> <span class="va">chdloddsse</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">17723</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cMLObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Constrained maximum likelihood method (MRcML) </span></span>
<span><span class="co">## Number of Variants:  28 </span></span>
<span><span class="co">## Results for:  cML-MA-BIC-DP </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##         Method Estimate    SE Pvalue        95% CI</span></span>
<span><span class="co">##  cML-MA-BIC-DP    2.821 0.388  0.000 [2.060,3.582]</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="principal-components-gmm-method">Principal components GMM method<a class="anchor" aria-label="anchor" href="#principal-components-gmm-method"></a>
</h3>
<p>The principal components GMM method is an implementation of the GMM
method with summarized data that is designed for use when performing
Mendelian randomization using genetic variants from a single gene
region. As an alternative to pruning and clumping approaches, which take
a large number of variants from a gene region (potentially hundreds or
thousands) and select a small number of uncorrelated (or weakly
correlated) variants, the principal components approach performs
dimension reduction on the full set of variant associations. The
principal components are then used as instrument variables, rather than
the individual genetic variants. The method is described in Patel et al
(2023) “Robust use of phenotypic heterogeneity at drug target genes for
mechanistic insights: application of cis-multivariable Mendelian
randomization to GLP1R gene region”, doi:
10.1101/2023.07.20.23292958.</p>
<p>Note that this method requires the sample size used to calculate the
genetic associations with the exposure (<em>nx</em>) and outcome
(<em>ny</em>) to be specified. The number of principal components
includes as instruments can either be set directly, by setting
<em>r</em>, or indirectly, by setting <em>thres</em>. The default option
<em>thres = 0.999</em> includes enough principal components to explain
99.9% of the variance in the weighted variant correlation matrix.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcGMMObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_pcgmm.html">mr_pcgmm</a></span><span class="op">(</span><span class="va">MRInputObject.cor</span>,</span>
<span>                        <span class="va">nx</span>,</span>
<span>                        <span class="va">ny</span>,</span>
<span>                        r <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        thres <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>                        robust <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcGMMObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_pcgmm.html">mr_pcgmm</a></span><span class="op">(</span><span class="fu"><a href="../reference/mr_input.html">mr_input</a></span><span class="op">(</span>bx <span class="op">=</span> <span class="va">calcium</span>, bxse <span class="op">=</span> <span class="va">calciumse</span>,</span>
<span>   by <span class="op">=</span> <span class="va">fastgluc</span>, byse <span class="op">=</span> <span class="va">fastglucse</span>, corr <span class="op">=</span> <span class="va">calc.rho</span><span class="op">)</span>,</span>
<span>   nx<span class="op">=</span><span class="fl">6351</span>, ny<span class="op">=</span><span class="fl">133010</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pcGMMObject</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Univariable principal components generalized method of moments (PC-GMM) method</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of principal components used : 5 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Robust model with overdispersion heterogeneity.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">##  Method Estimate Std Error 95% CI       p-value F-stat</span></span>
<span><span class="co">##  PC-GMM    2.435     0.786 0.895, 3.975   0.002   11.2</span></span>
<span><span class="co">## ------------------------------------------------------------------</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Overdispersion heterogeneity parameter estimate = -12.65081 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Heterogeneity test statistic = 1.6414</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="summaries-of-multiple-methods">Summaries of multiple methods<a class="anchor" aria-label="anchor" href="#summaries-of-multiple-methods"></a>
</h3>
<p>The <em>mr_allmethods</em> function is provided to easily compare
results (Estimate, Standard Error, 95% CI, and p-value) from multiple
methods. One can look at results from a wide range of methods
(<em>method = “all”</em>), or a limited set of results by setting method
to <em>“egger”</em>, <em>“ivw”</em>, or <em>“median”</em>. The final
option is <em>“main”</em>, which gives results from the simple median,
weighted median, IVW, and MR-Egger methods only.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRAllObject_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_allmethods.html">mr_allmethods</a></span><span class="op">(</span><span class="va">MRInputObject</span>, method <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="va">MRAllObject_all</span></span></code></pre></div>
<pre><code><span><span class="co">##                     Method Estimate Std Error 95% CI        P-value</span></span>
<span><span class="co">##              Simple median    1.755     0.740   0.305 3.205   0.018</span></span>
<span><span class="co">##            Weighted median    2.683     0.419   1.862 3.504   0.000</span></span>
<span><span class="co">##  Penalized weighted median    2.681     0.420   1.857 3.505   0.000</span></span>
<span><span class="co">##                                                                    </span></span>
<span><span class="co">##                        IVW    2.834     0.530   1.796 3.873   0.000</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 12 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRAllObject_egger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_allmethods.html">mr_allmethods</a></span><span class="op">(</span><span class="va">MRInputObject</span>, method <span class="op">=</span> <span class="st">"egger"</span><span class="op">)</span></span>
<span><span class="va">MRAllObject_egger</span></span></code></pre></div>
<pre><code><span><span class="co">##              Method Estimate Std Error 95% CI        P-value</span></span>
<span><span class="co">##            MR-Egger    3.253     0.770   1.743 4.762   0.000</span></span>
<span><span class="co">##         (intercept)   -0.011     0.015  -0.041 0.018   0.451</span></span>
<span><span class="co">##  Penalized MR-Egger    3.421     0.531   2.380 4.461   0.000</span></span>
<span><span class="co">##         (intercept)   -0.022     0.011  -0.043 0.000   0.051</span></span>
<span><span class="co">##     Robust MR-Egger    3.256     0.624   2.033 4.479   0.000</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 3 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MRAllObject_main</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr_allmethods.html">mr_allmethods</a></span><span class="op">(</span><span class="va">MRInputObject</span>, method <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span>
<span><span class="va">MRAllObject_main</span></span></code></pre></div>
<pre><code><span><span class="co">##           Method Estimate Std Error 95% CI        P-value</span></span>
<span><span class="co">##    Simple median    1.755     0.740   0.305 3.205   0.018</span></span>
<span><span class="co">##  Weighted median    2.683     0.419   1.862 3.504   0.000</span></span>
<span><span class="co">##              IVW    2.834     0.530   1.796 3.873   0.000</span></span>
<span><span class="co">##         MR-Egger    3.253     0.770   1.743 4.762   0.000</span></span>
<span><span class="co">##      (intercept)   -0.011     0.015  -0.041 0.018   0.451</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="final-note-of-caution">Final note of caution<a class="anchor" aria-label="anchor" href="#final-note-of-caution"></a>
</h3>
<p>Particularly with the development of Mendelian randomization with
summarized data, two-sample Mendelian randomization (where associations
with the risk factor and with the outcome are taken from separate
datasets), and bioinformatic tools for obtaining and analysing
summarized data (including this package), Mendelian randomization is
becoming increasingly accessible as a tool for answering epidemiological
questions. This is undoubtably a Good Thing.</p>
<p>However, it is important to remember that the difficult part of a
Mendelian randomization analysis is not the computational method, but
deciding what should go into the analysis: which risk factor, which
outcome, and (most importantly) which genetic variants. Hopefully, the
availability of these tools will enable less attention to be paid to the
mechanics of the analysis, and more attention to these choices.</p>
<p>The note of caution is that tools that make Mendelian randomization
simple to perform run the risk of encouraging large numbers of
speculative analyses to be performed in an unprincipled way. It is
important that Mendelian randomization is not performed in a way that
avoids critical thought. In releasing this package, the hope is that it
will lead to more comprehensive and more reproducible causal inferences
from Mendelian randomization, and not simply add more noise to the
literature.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Stephen Burgess, Olena Yavorska, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
